cpp/src/io/csv/writer_impl.cu

```diff
@@ -0,0 +1,488 @@
+                           stream));
+
+  CUDA_TRY(cudaStreamSynchronize(stream));
+
+  if (out_sink_->supports_device_write()) {
+    //host algorithm call, but the underlying call
+    //is a device_write taking a device buffer;
+    //
+    thrust::transform(thrust::host,
+                      h_offsets.begin(), h_offsets.end(),
+                      h_row_sizes.begin(),
+                      thrust::make_discard_iterator(),//discard output
+                      [&sink = out_sink_, ptr_all_bytes, stream] (size_type offset_indx, size_type row_sz) mutable {
+                        sink->device_write(ptr_all_bytes + offset_indx,
+                                           row_sz,
+                                           stream);
+                        return 0;//discarded (but necessary)
+                      });
+  } else {
+    //no device write possible;
+    //
+    //copy the bytes to host, too:
+    //
+    thrust::host_vector<char> h_bytes(total_num_bytes);
+    CUDA_TRY(cudaMemcpyAsync(h_bytes.data(), ptr_all_bytes, 
+                             total_num_bytes*sizeof(char),
+                             cudaMemcpyDeviceToHost,
+                             stream));
+
+    CUDA_TRY(cudaStreamSynchronize(stream));
+
+    //host algorithm call, where the underlying call
+    //is also host_write taking a host buffer;
+    //
+    char const* ptr_h_bytes = h_bytes.data();
+    thrust::transform(thrust::host,
+                      h_offsets.begin(), h_offsets.end(),
+                      h_row_sizes.begin(),
+                      thrust::make_discard_iterator(),//discard output
+                      [&sink = out_sink_, ptr_h_bytes, stream] (size_type offset_indx, size_type row_sz) mutable {
+                        sink->host_write(ptr_h_bytes + offset_indx,
+                                         row_sz);
+                        return 0;//discarded (but necessary)
+                      });
+  }
+}
+
+  
+void writer::impl::write(table_view const &table,
+                         const table_metadata *metadata,
+                         cudaStream_t stream) {
+  CUDF_EXPECTS( table.num_columns() > 0 && table.num_rows() > 0, "Empty table." );
+  
+
+  //write header: column names separated by delimiter:
+  //
+  write_chunked_begin(table, metadata, stream);
+  
+  //no need to check same-size columns constraint; auto-enforced by table_view
+  auto n_rows_per_chunk = options_.rows_per_chunk();
+  //
+  // This outputs the CSV in row chunks to save memory.
+  // Maybe we can use the total_rows*count calculation and a memory threshold
+  // instead of an arbitrary chunk count.
+  // The entire CSV chunk must fit in CPU memory before writing it out.
+  //
+  if( n_rows_per_chunk % 8 ) // must be divisible by 8
+    n_rows_per_chunk += 8 - (n_rows_per_chunk % 8);
+  CUDF_EXPECTS( n_rows_per_chunk>0, "write_csv: invalid chunk_rows; must be at least 8" );
```

Comment:
Why not assert what the message suggests?
```suggestion
  CUDF_EXPECTS( n_rows_per_chunk >= 8, "write_csv: invalid chunk_rows; must be at least 8" );
```

Reply:
Done.